prov/util/src/util_mr_cache.c

```diff
@@ -155,25 +155,20 @@ void ofi_mr_cache_notify(struct ofi_mr_cache *cache, const void *addr, size_t le
 
 bool ofi_mr_cache_flush(struct ofi_mr_cache *cache, bool flush_lru)
 {
+	bool	dead_region_result;
+	struct dlist_entry entries_to_free;
 	struct ofi_mr_entry *entry;
 
-	pthread_mutex_lock(&mm_lock);
-	while (!dlist_empty(&cache->flush_list)) {
-		dlist_pop_front(&cache->flush_list, struct ofi_mr_entry,
-				entry, list_entry);
-		FI_DBG(cache->domain->prov, FI_LOG_MR, "flush %p (len: %zu)\n",
-		       entry->info.iov.iov_base, entry->info.iov.iov_len);
-		pthread_mutex_unlock(&mm_lock);
-
-		util_mr_free_entry(cache, entry);
-		pthread_mutex_lock(&mm_lock);
-	}
+	dead_region_result = ofi_mr_cache_remove_dead_regions(cache);
```

Comment:
remove_dead_regions and flush implement the same logic and cleanup loop and should be combined.  Try something like:

```
dlist_init(free_list)
lock();
dlist_splice_tail(free_list, cache->dead_region_list)

while (flush_lru && !dlist_empty(lru_list) && cache_cnt < max_cnt ...)  {
    dlist_pop_front(lru_list. ...);
    util_mr_uncache_entry_storage(...);
    dlist_insert_tail(...);
}

unlock();
ret = !dlist_empty(free_list);
while (!dlist_empty(free_list) {
    dlist_pop_front(...)
    util_mr_free(...);
}

return ret;
```

Reply:
ok